tokio/src/io/async_read.rs

```diff
@@ -55,6 +85,1057 @@ pub trait AsyncRead {
+    ///
+    /// If this function encounters any form of I/O or other error, an error
+    /// variant will be returned. If an error is returned then it must be
+    /// guaranteed that no bytes were read.
+    ///
+    /// # Examples
+    ///
+    /// [`File`][crate::fs::File]s implement `Read`:
+    ///
+    /// ```no_run
+    /// use tokio::fs::File;
+    /// use tokio::io::{self, AsyncRead};
+    ///
+    /// #[tokio::main]
+    /// async fn main() -> io::Result<()> {
+    ///     let mut f = File::open("foo.txt").await?;
+    ///     let mut buffer = [0; 10];
+    ///
+    ///     // read up to 10 bytes
+    ///     let n = f.read(&mut buffer[..]).await?;
+    ///
+    ///     println!("The bytes: {:?}", &buffer[..n]);
+    ///     Ok(())
+    /// }
+    /// ```
+    fn read<'a>(&'a mut self, buf: &'a mut [u8]) -> Read<'a, Self>
+    where
+        Self: Unpin + Sized,
+    {
+        read(self, buf)
+    }
+
+    /// Pulls some bytes from this source into the specified buffer,
+    /// advancing the buffer's internal cursor.
+    ///
+    /// Equivalent to:
+    ///
+    /// ```ignore
+    /// async fn read_buf<B: BufMut>(&mut self, buf: &mut B) -> io::Result<usize>;
+    /// ```
+    ///
+    /// Usually, only a single `read` syscall is issued, even if there is
+    /// more space in the supplied buffer.
+    ///
+    /// This function does not provide any guarantees about whether it
+    /// completes immediately or asynchronously
+    ///
+    /// # Return
+    ///
+    /// On a successful read, the number of read bytes is returned. If the
+    /// supplied buffer is not empty and the function returns `Ok(0)` then
+    /// the source has reached an "end-of-file" event.
+    ///
+    /// # Errors
+    ///
+    /// If this function encounters any form of I/O or other error, an error
+    /// variant will be returned. If an error is returned then it must be
+    /// guaranteed that no bytes were read.
+    ///
+    /// # Examples
+    ///
+    /// [`File`] implements `Read` and [`BytesMut`] implements [`BufMut`]:
+    ///
+    /// [`File`]: crate::fs::File
+    /// [`BytesMut`]: bytes::BytesMut
+    /// [`BufMut`]: bytes::BufMut
+    ///
+    /// ```no_run
+    /// use tokio::fs::File;
+    /// use tokio::io::{self, AsyncRead};
+    ///
+    /// use bytes::BytesMut;
+    ///
+    /// #[tokio::main]
+    /// async fn main() -> io::Result<()> {
+    ///     let mut f = File::open("foo.txt").await?;
+    ///     let mut buffer = BytesMut::with_capacity(10);
+    ///
+    ///     assert!(buffer.is_empty());
+    ///
+    ///     // read up to 10 bytes, note that the return value is not needed
+    ///     // to access the data that was read as `buffer`'s internal
+    ///     // cursor is updated.
+    ///     f.read_buf(&mut buffer).await?;
+    ///
+    ///     println!("The bytes: {:?}", &buffer[..]);
+    ///     Ok(())
+    /// }
+    /// ```
+    fn read_buf<'a, B>(&'a mut self, buf: &'a mut B) -> async_read_futures::ReadBuf<'a, Self, B>
```

Comment:
@Darksonn @carllerche feel like we should rename this future?

Reply:
That would probably make sense @LucioFranco, but I'll leave it to the team to figure out what the name ought to be.

Reply:
The future is not public as of now. I would not worry about a rename yet. This is fine.